packages/kbn-observable/src/Subject.ts

```diff
@@ -1,114 +1,62 @@
+export class Subject<T> extends Observable<T> {
+  _observers: Set<SubscriptionObserver<T>> = new Set();
+  isStopped = false;
+  hasError = false;
+  thrownError?: Error;
 
   constructor() {
-    super(() => {});
+    super(observer => this.registerObserver(observer));
   }
 
-  next(value: T) {
-    if (this.closed) {
-      throw new ObjectUnsubscribedError();
-    }
-
+  protected registerObserver(observer: SubscriptionObserver<T>) {
     if (this.isStopped) {
-      return;
-    }
-
-    for (const observer of this.observers) {
-      observer.next!(value);
+      if (this.hasError) {
+        observer.error(this.thrownError!);
+      } else {
+        observer.complete();
+      }
+    } else {
+      this._observers.add(observer);
+      return () => this._observers.delete(observer);
     }
   }
 
-  error(err: any) {
-    if (this.closed) {
-      throw new ObjectUnsubscribedError();
+  next(value: T) {
+    const observers = Array.from(this._observers);
+    for (let observer of observers) {
+      observer.next(value);
     }
+  }
 
+  error(error: Error) {
     this.hasError = true;
-    this.thrownError = err;
+    this.thrownError = error;
     this.isStopped = true;
 
-    for (const observer of this.observers) {
-      observer.error!(err);
+    const observers = Array.from(this._observers);
+    for (let observer of observers) {
+      observer.error(error);
     }
-    this.observers.length = 0;
+
+    this._observers.clear();
   }
 
   complete() {
-    if (this.closed) {
-      throw new ObjectUnsubscribedError();
-    }
-
     this.isStopped = true;
 
-    for (const observer of this.observers) {
-      observer.complete!();
-    }
-    this.observers.length = 0;
-  }
-
-  // Subscribes to the sequence with an observer
-  subscribe(): Subscription;
-  subscribe(observer: PartialObserver<T>): Subscription;
-
-  // Subscribes to the sequence with callbacks
-  subscribe(
-    onNext: (val: T) => void,
-    onError?: (err: Error) => void,
-    onComplete?: () => void
-  ): Subscription;
-
-  subscribe(
-    observerOrOnNext?: PartialObserver<T> | ((value: T) => void),
-    onError?: (err: Error) => void,
-    onComplete?: () => void
-  ): Subscription {
-    const observer = toSubscriber(observerOrOnNext, onError, onComplete);
-    return this._subscribe(observer);
-  }
-
-  protected _subscribe(observer: Observer<T>): Subscription {
-    if (this.closed) {
-      throw new ObjectUnsubscribedError();
-    }
-
-    if (this.hasError) {
-      observer.error!(this.thrownError);
-      return SubjectSubscription.EMPTY;
+    const observers = Array.from(this._observers);
+    for (let observer of observers) {
+      observer.complete();
     }
 
-    if (this.isStopped) {
-      observer.complete!();
-      return SubjectSubscription.EMPTY;
-    }
-
-    this.observers.push(observer);
-
-    return new SubjectSubscription(() => {
-      this.observers.splice(this.observers.indexOf(observer), 1);
-    });
-  }
-
-  unsubscribe() {
```

Comment:
I removed `unsubscribe` for now. Not sure if it's ever needed in our context. In our usage we've just `complete`d it instead.